lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost/Layer.hs

```diff
@@ -0,0 +1,260 @@
+    , bfGetLatestEpochProtocolParams = getLatestEpochProtocolParams
+    , bfGetPoolHistory = getPoolHistory'
+    , bfGetLedgerGenesis = getLedgerGenesis
+    , bfGetAccount = maybe404 . getAccount
+    , bfGetBlockSlot = getBlockSlot
+    , bfGetBlockAtHeight = getBlock . Left
+    , bfGetBlockAfterHash =
+        consensual404 . (listToMaybe <$>) . (`getNextBlocks'` paged 1 1) . Right
+    , bfGetBlocksAfterHash =
+        consensual404 . allPages' . getNextBlocks' . Right
+    , bfGetBlockTxs = \a -> allPages' \p -> getBlockTxs' (Right a) p Ascending
+    , bfGetTx = getTx
+    , bfGetTxStakes = getTxStakes
+    , bfGetTxUtxos = getTxUtxos
+    , bfGetTxWithdrawals = getTxWithdrawals
+    , bfGetTxDelegations = getTxDelegations
+    , bfGetTxMetadataJSON = getTxMetadataJSON
+    , bfGetAddressTransactions = \a f t ->
+        empty404 $ allPages' \p -> getAddressTransactions' a p Ascending f t
+    , bfGetAccountRegistrations = \a ->
+        empty404 $ allPages' \p -> getAccountRegistrations' a p Ascending
+    , bfGetAccountDelegations = \a ->
+        empty404 $ allPages' \p -> getAccountDelegations' a p Ascending
+    , bfGetAccountWithdrawals = \a ->
+        empty404 $ allPages' \p -> getAccountWithdrawals' a p Ascending
+    , bfGetNetworkInfo = getNetworkInfo
+    , bfListPools = allPages' (`listPools'` Ascending)
+    }
+
+hoistBlockfrostLayer ::
+    BlockfrostLayer BFM -> (forall a. BFM a -> IO a) -> BlockfrostLayer IO
+hoistBlockfrostLayer BlockfrostLayer{..} nt =
+    BlockfrostLayer
+    { bfGetLatestBlock = nt bfGetLatestBlock
+    , bfGetLatestEpoch = nt bfGetLatestEpoch
+    , bfGetLatestEpochProtocolParams = nt bfGetLatestEpochProtocolParams
+    , bfGetPoolHistory = ((nt .) .) . bfGetPoolHistory
+    , bfGetLedgerGenesis = nt bfGetLedgerGenesis
+    , bfGetAccount = nt . bfGetAccount
+    , bfGetBlockSlot = nt . bfGetBlockSlot
+    , bfGetBlockAtHeight = nt . bfGetBlockAtHeight
+    , bfGetBlockAfterHash = nt . bfGetBlockAfterHash
+    , bfGetBlocksAfterHash = nt . bfGetBlocksAfterHash
+    , bfGetBlockTxs = nt . bfGetBlockTxs
+    , bfGetTx = nt . bfGetTx
+    , bfGetTxStakes = nt . bfGetTxStakes
+    , bfGetTxUtxos = nt . bfGetTxUtxos
+    , bfGetTxWithdrawals = nt . bfGetTxWithdrawals
+    , bfGetTxDelegations = nt . bfGetTxDelegations
+    , bfGetTxMetadataJSON = nt . bfGetTxMetadataJSON
+    , bfGetAddressTransactions = ((nt .) .) . bfGetAddressTransactions
+    , bfGetAccountRegistrations = nt . bfGetAccountRegistrations
+    , bfGetAccountDelegations = nt . bfGetAccountDelegations
+    , bfGetAccountWithdrawals = nt . bfGetAccountWithdrawals
+    , bfGetNetworkInfo = nt bfGetNetworkInfo
+    , bfListPools = nt bfListPools
+    }
+
+rateLimitedBlockfrostLayer :: Tracer IO Log -> Project -> IO (BlockfrostLayer IO)
+rateLimitedBlockfrostLayer tr project = do
+    bfConfig <- BFM.newClientConfig project
+    r <- newIORef . (,rps) =<< getPOSIXTime
+    pure $ hoistBlockfrostLayer blockfrostLayer \bfm -> do
```

Comment:
I 😍 Haskell. Straightforward boilerplate + 10 lines of logic for a rate limiter that just works.™️